### PR TITLE
macOS universal builds revisited

### LINKED ISSUE #408: Improve universal support for macOS
The current macOS support only handles a single x86 build slice, but it should be possible to support a single Universal binary that supports arm64 (NEON), x86_64 (SSE4.1), and x86_64h (AVX2) slices. 

This was attempted in PR #405, but the current build doesn't really cope with it as we try to build all slices in a single compile. Given the need for different compile options for each x86 slice we really need to compile multiple binaries with a custom universal build step that uses `lipo` to merge the requested slices into a single universal binary.

### PR DESCRIPTION
**Pull request by solidpixel**
This PR provides a new implementation of macOS universal binary builds, which is considerably simpler than the previous integration and more useful in as far as it supports the `x86_64h` slice for AVX2. 

This build repurposes the existing `ASTCENC_UNIVERSAL_BUILD` option. When enabled (and it's on by default for macOS) this will automatically build the SSE41, AVX2 and NEON single ISA build variants, and use `lipo` to package these three together into a single binary. When in universal mode only the combined binary is installed; the single ISA binaries are built but are treated as internal intermediate build artefacts.

This is the only ISA support provided by the universal build - you can't build with a subset, and you can't change to use NONE or SSE2 in the `x86_64` slice or NONE in the `arm64` slice. All ISA options are still available as single architecture builds.  

The Jenkins pipelines for the macOS binary releases have been updated to release a single universal binary, instead of separate builds per ISA, and separate bundles for x86_64 and arm64.

Fixes #408

### CODE REVIEW

File: Source/CMakeLists.txt [hunk 1]
```diff
@@ -27,35 +27,70 @@ else()
...
+
+    install(PROGRAMS ${CMAKE_CURRENT_BINARY_DIR}/libastc${ASTCENC_CODEC}-shared.dylib
+            DESTINATION lib)
 endif()
 
 # - - - - - - - - - - - - - - - - - -
 # Unit testing
 if(${ASTCENC_UNITTEST})
     set(INSTALL_GTEST OFF CACHE BOOL "" FORCE)
+    set(CMAKE_OSX_ARCHITECTURES x86_64;arm64)
```
Comment (wasimabbas-arm) [inline Source/CMakeLists.txt:93]: Should there be `x86_64h` target here as well for testing?

**Review by wasimabbas-arm — CHANGES_REQUESTED** [review]

**Comment by solidpixel** [comment]
@wasimabbas-arm reports that the XCode builder is currently failing: 

```cmake -H. -G Xcode -Bbuild-metal-debug-xcode -DCMAKE_BUILD_TYPE=Release && cmake --build build-metal-debug-xcode --config Release```

We should also add XCode build instructions to the Markdown - it's not something I've tried before.

File: CMakeLists.txt [hunk 2]
```diff
@@ -75,104 +88,8 @@ foreach(ASTCENC_CONFIG ${ASTCENC_CONFIGS})
...
-    else()
-        # Do nothing here - assume it defaults to host?
-
-    endif()
-
 # Non-macOS builds
-else()
+if(NOT "${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
     if(("${ASTCENC_ARM64_ISA_COUNT}" GREATER 0) AND ("${ASTCENC_X64_ISA_COUNT}" GREATER 0))
         message(FATAL_ERROR "Builds can only support a single architecture per configure.")
```
[the remark this answers is not in the document]
Comment (wasimabbas-arm) [reply CMakeLists.txt:94]: `-DASTCENC_ISA_NEON=ON -DASTCENC_ISA_AVX2=ON` that without the `-DASTCENC_UNIVERSAL_BUILD=OFF` does build fine.

**Comment by wasimabbas-arm** [comment]
The last two commits seems to have fixed the Xcode build.

**Comment by solidpixel** [comment]
Confirmed that CI builds all pass for this now, on all three operating systems.

**Pull request merged**